### PR TITLE
Add design doc for DeepSpeech2 on PaddlePaddle.

### PR DESCRIPTION
**Pull request by xinghai-sun**
This is just a simple and incomplete version, with parts to be added:
- CTC-LM beam search decoder (important!)
- Lookahead row convolution

### CODE REVIEW

File: doc/design/speech/README.MD [hunk 1]
```diff
@@ -0,0 +1,155 @@
+# DeepSpeech2 on PaddlePaddle: Design Doc 
+
+We are planning to build Deep Speech 2 (DS2) \[[1](#references)\], a powerful Automatic Speech Recognition (ASR) engine,  on PaddlePaddle. For the first-stage plan, we have the following short-term goals:
+
+- Release a basic distributed implementation of DS2 on PaddlePaddle.
+- Contribute a chapter of Deep Speech to PaddlePaddle Book.
```
Comment (wangkuiyi) [inline doc/design/speech/README.MD]: Do we also need the (not-yet-optimized) inference part?

File: doc/design/speech/README.MD [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	- Audio file format transformer.
+	- Spectrogram feature extraction, power normalization etc.
+	- Batch data reader with SortaGrad.
+	- Data augmentation (optional).
+	- Prepare (one or more) public English data sets & baseline.
+2. Create a **simplified DS2 model configuration**:
+   - With only fixed-length (by padding) audio sequences (otherwise need *Task 3*).
+	- With only bidirectional-GRU (otherwise need *Task 4*).
+	- With only greedy decoder (otherwise need *Task 5, 6*).
+3. Develop to support **variable-shaped** dense-vector (image) batches of input data.
```
Comment (wangkuiyi) [inline doc/design/speech/README.MD]: Is this related to the development of Tensor?  @xinghai-sun might want to talk to @QiJune @Canpio and @gangliao .

File: doc/design/speech/README.MD [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    - [Row Convolution](#row-convolution)
+    - [Beam Search With CTC and LM](#beam-search-with-ctc-and-lm)
+- [Future Work](#future-work)
+- [References](#references)
+
+## Tasks
+
+We roughly break down the project into 14 tasks:
+
+1. Develop an **audio data provider**:
```
Comment (wangkuiyi) [inline doc/design/speech/README.MD]: Would it help if AI Lab provides us their Python implementation of their audio data preprocessor?

File: doc/design/speech/README.MD [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+   - update `ctc_error_evaluator`(CER) to support WER.
+8. Prepare internal dataset for Mandarin (optional):
+    - Dataset, baseline, evaluation details.
+    - Particular data preprocessing for Mandarin.
+    - Might need cooperating with the Speech Department.
+9. Create **standard DS2 model configuration**:
+   - With variable-length audio sequences (need *Task 3*).
+	- With unidirectional-GRU + row-convolution (need *Task 4*).
+	- With CTC-LM beam search decoder (need *Task 5, 6*).
+10. Make it run perfectly on **clusters**.
```
Comment (wangkuiyi) [inline doc/design/speech/README.MD]: Do we mean MPI clusters or Kubernetes here?

File: doc/design/speech/README.MD [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+11. Experiments and **benchmarking** (for accuracy, not efficiency):
+    - With public English dataset.
+    - With internal (Baidu) Mandarin dataset (optional).
```
Comment (wangkuiyi) [inline doc/design/speech/README.MD]: I heart that Baidu is going to publish a Mandarin dataset. Do you know if this is true, and if so, when will they do so?

File: doc/design/speech/README.MD (hunk 2 above)
Comment (xinghai-sun) [inline doc/design/speech/README.MD]: Yes, Tensor would make it more flexible. Currently, @qingqing01 has added a new data type to temporarily meet the need. Will talk to them later. Looking forward to the coming of Tensor!

File: doc/design/speech/README.MD (hunk 5 above)
Comment (xinghai-sun) [inline doc/design/speech/README.MD]: I don't know about it. @lcy-seso Could you help collect the information?

File: doc/design/speech/README.MD (hunk 4 above)
Comment (xinghai-sun) [inline doc/design/speech/README.MD]: I hope for both.
